src/types.jl

```diff
@@ -56,6 +56,55 @@ symtype(x) = typeof(x)
 
 symtype(::Symbolic{T}) where {T} = T
 
+function hasmetadata(s::Symbolic, ctx)
+    s.metadata isa Ref && haskey(s.metadata[], ctx)
+end
+
+function getmetadata(s::Symbolic, ctx)
+    if s.metadata isa Ref
+        s.metadata[][ctx]
+    else
+        throw(ArgumentError("$s does not have metadata for $ctx"))
+    end
+end
+
+function getmetadata(s::Symbolic, ctx, default)
+    s.metadata isa Ref ? get(s.metadata[], ctx, default) : default
+end
+
+import Base: ImmutableDict
+import Setfield: PropertyLens
+
+# pirated for Setfield purposes:
+Base.ImmutableDict(d::ImmutableDict{K,V}, x, y)  where {K, V} = ImmutableDict{K,V}(d, x, y)
+
+assoc(d::Dict, ctx, val) = (d=copy(d); d[ctx] = val; d)
+function assoc(d::Base.ImmutableDict, ctx, val)::ImmutableDict{DataType,Any}
+    # optimizations
+    # If using upto 3 contexts, things stay compact
+    if isdefined(d, :parent)
+        d.key === ctx && return @set d.value = val
+        d1 = d.parent
+    end
+    if isdefined(d1, :parent)
+        d1.key === ctx && return @set d.parent.value = val
+        d2 = d1.parent
+    end
+    if isdefined(d2, :parent)
+        d2.key === ctx && return @set d.parent.parent.value = val
+    end
+    Base.ImmutableDict{DataType, Any}(d, ctx, val)
+end
+
+function setmetadata(s::Symbolic, ctx::DataType, val)
+    if s.metadata isa Ref
+        @set s.metadata = Ref{Any}(assoc(s.metadata[], ctx, val))
+    else
+        # fresh Dict
+        @set s.metadata = Ref{Any}(Base.ImmutableDict{DataType, Any}(ctx, val))
```

Comment:
@YingboMa I think this Ref is not necessary here, it will be `Union{Nothing, ImmutableDict{DataType, Any}}`.

Reply:
Just as good. In fact we can even use `ImmutableDict{DataType, Any}()` as the default "empty" thing, which would mean no type parameter. But that empty struct adds 24 bytes to Sym... Which maybe just fine!

Reply:
Have you benchmarked it?

Reply:
No I didn't benchmark the empty ImmutableDict, right now it is still a parameter which can be Nothing or ImmutableDict.

Reply:
This is faster than having a Ref of ImmutableDict.